model/atmosphere/dycore/src/icon4py/model/atmosphere/dycore/stencils/compute_horizontal_velocity_quantities.py

```diff
@@ -0,0 +1,269 @@
+    rbf_vec_coeff_e: gtx.Field[gtx.Dims[dims.EdgeDim, dims.E2C2EDim], ta.wpfloat],
+    rho_at_edges_on_model_levels: fa.EdgeKField[ta.wpfloat],
+    theta_v_at_edges_on_model_levels: fa.EdgeKField[ta.wpfloat],
+    ddqz_z_full_e: fa.EdgeKField[ta.vpfloat],
+    ddxn_z_full: fa.EdgeKField[ta.vpfloat],
+    ddxt_z_full: fa.EdgeKField[ta.vpfloat],
+    wgtfac_e: fa.EdgeKField[ta.vpfloat],
+    wgtfacq_e: fa.EdgeKField[ta.vpfloat],
+    nflatlev: gtx.int32,
+    horizontal_start: gtx.int32,
+    horizontal_end: gtx.int32,
+    vertical_start: gtx.int32,
+    vertical_end: gtx.int32,
+):
+    _compute_horizontal_velocity_quantities_and_fluxes(
+        contravariant_correction_at_edges_on_model_levels,
+        vn,
+        e_flx_avg,
+        geofac_grdiv,
+        rbf_vec_coeff_e,
+        rho_at_edges_on_model_levels,
+        theta_v_at_edges_on_model_levels,
+        ddqz_z_full_e,
+        ddxn_z_full,
+        ddxt_z_full,
+        wgtfac_e,
+        nflatlev,
+        out=(
+            spatially_averaged_vn,
+            horizontal_gradient_of_normal_wind_divergence,
+            tangential_wind,
+            mass_flux_at_edges_on_model_levels,
+            theta_v_flux_at_edges_on_model_levels,
+            vn_on_half_levels,
+            tangential_wind_on_half_levels,
+            horizontal_kinetic_energy_at_edges_on_model_levels,
+            contravariant_correction_at_edges_on_model_levels,
+        ),
+        domain={
+            dims.EdgeDim: (horizontal_start, horizontal_end),
+            dims.KDim: (vertical_start, vertical_end - 1),
+        },
+    )
+
+    _extrapolate_at_top(
+        wgtfacq_e,
+        vn,
+        out=vn_on_half_levels,
+        domain={
+            dims.EdgeDim: (horizontal_start, horizontal_end),
+            dims.KDim: (vertical_end - 1, vertical_end),
+        },
+    )
+
+
+@gtx.field_operator
+def _compute_averaged_vn_and_fluxes_and_prepare_tracer_advection(
+    substep_and_spatially_averaged_vn: fa.EdgeKField[ta.wpfloat],
+    substep_averaged_mass_flux: fa.EdgeKField[ta.wpfloat],
+    e_flx_avg: gtx.Field[gtx.Dims[dims.EdgeDim, dims.E2C2EODim], ta.wpfloat],
+    vn: fa.EdgeKField[ta.wpfloat],
+    rho_at_edges_on_model_levels: fa.EdgeKField[ta.wpfloat],
+    ddqz_z_full_e: fa.EdgeKField[ta.vpfloat],
+    theta_v_at_edges_on_model_levels: fa.EdgeKField[ta.wpfloat],
+    prepare_advection: bool,
+    at_first_substep: bool,
+    r_nsubsteps: ta.wpfloat,
+) -> tuple[
+    fa.EdgeKField[ta.wpfloat],
+    fa.EdgeKField[ta.wpfloat],
+    fa.EdgeKField[ta.wpfloat],
+    fa.EdgeKField[ta.wpfloat],
+    fa.EdgeKField[ta.wpfloat],
+]:
+    spatially_averaged_vn = _compute_avg_vn(e_flx_avg, vn)
+
+    (
+        mass_flux_at_edges_on_model_levels,
+        theta_v_flux_at_edges_on_model_levels,
+    ) = _compute_mass_and_temperature_flux(
+        rho_at_edges_on_model_levels,
+        spatially_averaged_vn,
+        ddqz_z_full_e,
+        theta_v_at_edges_on_model_levels,
+    )
```

Comment:
from a readability point of view the function should probably be split: compute_mass_flux and compute_temperature_flux (or since the field_operator looks trivial, maybe inlined?)

Reply:
I think it is nice to compute both fluxes together (basically as a domain scientist in ICON you know that you need to compute both fluxes, and both fluxes need to be handed to the tracer advection.
Also the temperature flux directly depends on the mass flux, as you see in the formula
```
    mass_fl_e_wp = z_rho_e * z_vn_avg * astype(ddqz_z_full_e, wpfloat)
    z_theta_v_fl_e_wp = mass_fl_e_wp * z_theta_v_e
```
So I would leave it like this and let the domain scientist change it if they want to do a bigger re-organization.